src/kbc_modules/eaa_kbc/mod.rs

```diff
@@ -0,0 +1,223 @@
+        }
+    }
+
+    fn kbs_decrypt_payload(&mut self) -> Result<Vec<u8>> {
+        let request = DecryptionRequest {
+            command: String::from("Decrypt"),
+            blobs: vec![Blob {
+                kid: self.key_id.clone(),
+                encrypted_data: base64::encode(&self.encrypted_payload),
+                algorithm: "AES".to_string(),
+                key_length: 256,
+                iv: base64::encode(&self.iv),
+            }],
+        };
+        let trans_json = serde_json::to_string(&request)?;
+        let trans_data: &[u8] = trans_json.as_bytes();
+
+        let reciv_string: String = self.kbs_trans_and_reciv(trans_data, "Dcryption")?;
+        let response: DecryptionResponse =
+            serde_json::from_str::<DecryptionResponse>(reciv_string.as_str())?;
+
+        let payload_hashmap = match response.status.as_str() {
+            "OK" => response.data,
+            "Fail" => return Err(anyhow!(format!("Verdictd: {}", response.error.unwrap()))),
+            _ => return Err(anyhow!("Can't understand the DcryptionResponse status!")),
+        };
+
+        if let Some(hashmap_content) = payload_hashmap {
+            let encrypted_payload_string = base64::encode(&self.encrypted_payload);
+            let decrypted_payload_string = hashmap_content.get(&encrypted_payload_string).ok_or(anyhow!(
+                "There is no field matching the encrypted payload in the data field of DcryptionResponse"
+            ))?;
+            let decrypted_payload = base64::decode(decrypted_payload_string)?;
+            return Ok(decrypted_payload);
+        } else {
+            return Err(anyhow!(
+                "DecryptionResponse status is OK but the data is null!"
+            ));
+        }
+    }
+
+    fn kbs_get_kek(&mut self) -> Result<()> {
+        let request = GetKekRequest {
+            command: String::from("Get KEK"),
+            kids: vec![self.key_id.clone()],
+        };
+        let trans_json = serde_json::to_string(&request)?;
+        let trans_data: &[u8] = trans_json.as_bytes();
+
+        let reciv_string: String = self.kbs_trans_and_reciv(trans_data, "Get KEK")?;
+        let response: GetKekResponse =
+            serde_json::from_str::<GetKekResponse>(reciv_string.as_str())?;
+
+        let kek_hashmap = match response.status.as_str() {
+            "OK" => response.data,
+            "Fail" => return Err(anyhow!(format!("Verdictd: {}", response.error.unwrap()))),
+            _ => return Err(anyhow!("Can't understand the GetKekResponse status!")),
+        };
+
+        if let Some(hashmap_content) = kek_hashmap {
+            for (kid, kek_string) in &hashmap_content {
+                let kek = base64::decode(kek_string)?;
+                self.kek_cache.insert(kid.to_string(), kek);
+            }
+        } else {
+            return Err(anyhow!("GetKekResponse status is OK but the key is null!"));
+        }
+        Ok(())
```

Comment:
If it were me, I would have this fucntion return the KEK rather than just update the cache. The first thing the function would do is check the kek_cache. If the requested kid is in the cache it would immediately return it. Currently it seems like the calling function has to take care of that.